### DISCUSSION TITLE
Reporting actual storage allocated by dedup enabled zvol.

### DISCUSSION THREAD

**Discussion by datacore-rm** [openzfsonwindows/openzfs#342] [General]
Question is similar to https://github.com/openzfs/zfs/issues/14664.

We are trying to report storage savings per zvol on using zfs. For dedup enabled zvols, the 'USED' column in the _zfs list_ cmd shows the referenced space and not the actual storage space used. The _zpool status -D_ shows the pool wide actual dedup allocated space. When there are multiple dedup enabled zvols, is there any way to report the actual storage used by each zvol?

E.g The two zvols which are showing referenced space 30G & 20G below are dedup enabled zvols. The actual storage used is 12.5G. Both zvols could be referencing many same DDT entries.

```
C:\zfs>zpool list
NAME                                     SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
Z.0100cf19-f8c3-4b0f-a5dc-274a74cf3612    49G  12.6G  36.4G        -         -     1%    25%  3.99x    ONLINE  -

C:\zfs>zfs list
NAME                                                                          USED  AVAIL     REFER  MOUNTPOINT
Z.0100cf19-f8c3-4b0f-a5dc-274a74cf3612                                       50.1G  34.9G       96K  none
Z.0100cf19-f8c3-4b0f-a5dc-274a74cf3612/168f8d33-194f-44bd-8bbb-7ec42347b206   108K  34.9G      108K  -
Z.0100cf19-f8c3-4b0f-a5dc-274a74cf3612/758418d1-0044-4c26-8a0b-818502819495  20.0G  34.9G     20.0G  -
Z.0100cf19-f8c3-4b0f-a5dc-274a74cf3612/7aa3fe2d-6f53-4198-a190-4fcdc605104d  30.0G  34.9G     30.0G  -
Z.0100cf19-f8c3-4b0f-a5dc-274a74cf3612/9dc48e93-b71c-4ed3-921c-649a20607105   108K  34.9G      108K  -
Z.0100cf19-f8c3-4b0f-a5dc-274a74cf3612/c527e4aa-4d38-4b30-9f5b-14c7735ac253   108K  34.9G      108K  -
Z.0100cf19-f8c3-4b0f-a5dc-274a74cf3612/faed2cd4-cfba-43c4-a21a-a360046ca310   108K  34.9G      108K  -

C:\zfs>zpool status -D
  pool: Z.0100cf19-f8c3-4b0f-a5dc-274a74cf3612
 state: ONLINE
config:

        NAME                                    STATE     READ WRITE CKSUM
        Z.0100cf19-f8c3-4b0f-a5dc-274a74cf3612  ONLINE       0     0     0
          physicaldrive1                        ONLINE       0     0     0
          physicaldrive2                        ONLINE       0     0     0

errors: No known data errors

 dedup: DDT entries 102417, size 524B on disk, 169B in core

bucket              allocated                       referenced
______   ______________________________   ______________________________
refcnt   blocks   LSIZE   PSIZE   DSIZE   blocks   LSIZE   PSIZE   DSIZE
------   ------   -----   -----   -----   ------   -----   -----   -----
     1    80.0K   10.0G   10.0G   10.0G    80.0K   10.0G   10.0G   10.0G
     8    7.49K    958M    958M    958M     105K   13.2G   13.2G   13.2G
    16    12.5K   1.56G   1.56G   1.56G     215K   26.8G   26.8G   26.8G
 Total     100K   12.5G   12.5G   12.5G     400K   50.0G   50.0G   50.0G
```

Thanks

**Comment by lundman** [comment]
From slack:



Rich Ercolani
  [08:39](https://openzfs.slack.com/archives/C052RGXL5/p1705621190076009)
I believe the answer directly is "no", unless you go do a secondary lookup to figure out if it's in the "unique" or "not unique" DDT atm.
[08:39](https://openzfs.slack.com/archives/C052RGXL5/p1705621199223899)
Per block.
[08:40](https://openzfs.slack.com/archives/C052RGXL5/p1705621235620349)
Allan et al have been in the dedup code recently, so maybe I'm wrong and they'll know better, but I don't believe there's any kind of granular accounting of the type you'd want to not have to pay that price to do the math.


Rich Ercolani
  [09:46](https://openzfs.slack.com/archives/C052RGXL5/p1705625169757249)
I think the problem would be that you'd have to go recalculate it anytime you moved a block between DDTs, and that would, ahem, hurt.


Jörgen Lundman
  [10:03](https://openzfs.slack.com/archives/C052RGXL5/p1705626224243729)
Ah yeah that would hurt


Rich Ercolani
  [10:28](https://openzfs.slack.com/archives/C052RGXL5/p1705627718987109)
Seeing what percent of blocks are in the DDT would be ~free since it's just a bit on the records, but knowing how much is uniquely used is a bit more expensive.
[10:29](https://openzfs.slack.com/archives/C052RGXL5/p1705627761934169)
I realize I don't know if destroy -nv accounts for space savings from dedup in its estimates. I would guess not, since the way we present space savings in dedup/BRT is USED and TOTAL both increase.
[10:30](https://openzfs.slack.com/archives/C052RGXL5/p1705627806952749)
But if it does, you could steal however the guts of that decides if a block is uniquely referenced, potentially.
[10:30](https://openzfs.slack.com/archives/C052RGXL5/p1705627827039039)
(Not that the check is hard, or even expensive, if the DDT is in RAM. If it's not...well.)


Rob N
  [14:40](https://openzfs.slack.com/archives/C052RGXL5/p1705642827435909)
yeah, you can't do this at all. dedup table is (conceptually) a block checksum -> (dva[], refcount) mapping. there's no knowledge of  datasets or any other "structure" like that; its just blocks


Rich Ercolani
  [18:30](https://openzfs.slack.com/archives/C052RGXL5/p1705656608325229)
I mean, you can, it's just going to be an expensive calculation every time you ask, unless you did something like throttling how often it updates.
[18:31](https://openzfs.slack.com/archives/C052RGXL5/p1705656705829709)
And since I think the userquota stuff doesn't necessarily promise to be always up to date, particularly in cases where it's recalculating, maybe that's okay?
[18:32](https://openzfs.slack.com/archives/C052RGXL5/p1705656722866829)
(Just as an analogue of a property that's not always realtime accurate.)

Allan Jude
  [04:26](https://openzfs.slack.com/archives/C052RGXL5/p1705692370280539)
I think the issue is, how do you decide which zvol "owns" any particular block

**Reply by datacore-rm** [reply]
Thanks @lundman for sharing this info.
